config/main.py

```diff
@@ -2064,6 +2064,538 @@ def delete_snmptrap_server(ctx, ver):
+                   "symbols {}".format(snmp_username, excluded_special_symbols))
+        return False
+    return True
+
+
+def snmp_user_secret_check(snmp_secret):
+    excluded_special_symbols = ['@', ":"]
+    if len(snmp_secret) < 8:
+        click.echo("SNMP user password length should be at least 8 characters")
+        click.echo("SNMP user password length should be not be greater than 64")
+        click.echo("SNMP user password should not have any of these special "
+                   "symbols {}".format(excluded_special_symbols))
+        click.echo("FAILED: SNMP user password length should be at least 8 characters")
+        return False
+    if len(snmp_secret) > 64:
+        click.echo("SNMP user password length should be at least 8 characters")
+        click.echo("SNMP user password length should be not be greater than 64")
+        click.echo("SNMP user password should not have any of these special "
+                   "symbols {}".format(excluded_special_symbols))
+        click.echo("FAILED: SNMP user password length should be not be greater than 64")
+        return False
+    if any(char in excluded_special_symbols for char in snmp_secret):
+        click.echo("SNMP user password length should be at least 8 characters")
+        click.echo("SNMP user password length should be not be greater than 64")
+        click.echo("SNMP user password should not have any of these special "
+                   "symbols {}".format(excluded_special_symbols))
+        click.echo("FAILED: SNMP user password should not have any of these special "
+                   "symbols {}".format(excluded_special_symbols))
+        return False
+    return True
+
+
+@community.command('add')
+@click.argument('community', metavar='<snmp_community>', required=True)
+@click.argument('string_type', metavar='<RO|RW>', required=True)
+@clicommon.pass_db
+def add_community(db, community, string_type):
+    """ Add snmp community string"""
+    string_type = string_type.upper()
+    if not is_valid_community_type(string_type):
+        click.echo("SNMP type must be RO or RW and not {}".format(string_type))
+        sys.exit(1)
+    if not snmp_community_secret_check(community):
+        click.echo("Bad community")
+        sys.exit(2)
+    snmp_communities = db.cfgdb.get_table("SNMP_COMMUNITY")
+    if community in snmp_communities:
+        click.echo("SNMP community {} is already configured".format(community))
+        sys.exit(3)
+    else:
+        db.cfgdb.set_entry('SNMP_COMMUNITY', community, {'TYPE': string_type})
+        click.echo("SNMP community {} added to configuration".format(community))
+        try:
+            click.echo("Restarting SNMP service...")
+            clicommon.run_command("systemctl reset-failed snmp.service", display_cmd=False)
+            clicommon.run_command("systemctl restart snmp.service", display_cmd=False)
+        except SystemExit as e:
+            db.cfgdb.set_entry('SNMP_COMMUNITY', community, None)
```

Comment:
>set_entry [](http://example.com/codeflow?start=21&length=9)

Why you set SNMP_COMMUNITY again? It's dangerous to execute complex operation inside `except` block.

Reply:
If this failed when trying to set SNMP_COMMUNITY originally wouldn't it be better to try to remove it?  Or you would rather this just exit and not retry this?

Reply:
I understand you are trying to recover. However set SNMP_COMMUNITY to None is not enough for a complete recovery, and it may confuse user when troubleshoot. Prefer exit directly.

Reply:
ok removed.